File: Client/Captions/CaptionsLib/source/utils/DateTimeUtils.vb
'Filename: DateTimeUtils.vb
'Version: 20121016

Namespace ClipFlair.CaptionsLib.Utils

  Public NotInheritable Class DateTimeUtils

#Region "Constants"

    Public Shared ReadOnly DATETIMEZERO As New DateTime(1, 1, 1, 0, 0, 0, DateTimeKind.Local) '01/01/0001, 0:0:0.00 (dates must be from year 01 and above) '??? trying bigger year to avoid accuracy errors but it causes hichkups at the timebar

#End Region

#Region "Methods"

    Public Shared Function FixDateTime(ByVal datetime As DateTime) As DateTime
      With DATETIMEZERO
        Return New DateTime(.Year, .Month, .Day, datetime.Hour, datetime.Minute, datetime.Second, datetime.Millisecond)
      End With
    End Function

    Public Shared Function DateTimeToSeconds(ByVal theDateTime As DateTime, ByVal baseDateTime As DateTime, ByVal digits As Integer) As Double
      Dim result As Double = (theDateTime - baseDateTime).TotalSeconds
      If digits <> -1 Then Return Math.Round(result, digits)
      Return result
    End Function

    Public Shared Function SecondsToDateTime(ByVal seconds As Double, ByVal baseDateTime As DateTime, ByVal digits As Integer) As DateTime
      Try
        If digits <> -1 Then seconds = Math.Round(seconds, digits)
        Return baseDateTime.AddSeconds(seconds)
      Catch
        Return baseDateTime 'silently handle exceptions, returning the baseDateTime in that case
      End Try
    End Function

    Public Shared Function TimeStrToDateTime(ByVal datetimeStr As String, ByVal baseDateTime As DateTime, ByVal datetimeFormat As String) As DateTime
      Dim result As DateTime
      Try
        result = DateTime.ParseExact(datetimeStr, datetimeFormat, Nothing)
      Catch 'if exact datetime parsing fails, try with the plain parse method
        result = DateTime.Parse(datetimeStr)
      End Try
      result = FixDateTime(baseDateTime + (result - baseDateTime))
      Return result
    End Function

    Public Shared Function TimeStrToSeconds(ByVal datetimeStr As String, ByVal baseDateTime As DateTime, ByVal datetimeFormat As String, ByVal digits As Integer) As Double
      Return DateTimeToSeconds(TimeStrToDateTime(datetimeStr, baseDateTime, datetimeFormat), baseDateTime, digits)
    End Function

    Public Shared Function SecondsToDateTimeStr(ByVal seconds As Double, ByVal baseDateTime As DateTime, ByVal datetimeFormat As String, ByVal digits As Integer) As String
      Return SecondsToDateTime(seconds, baseDateTime, digits).ToString(datetimeFormat)
    End Function

#End Region

  End Class

End Namespace

File: Client/Media/Captions/CaptionsLib/source/utils/DateTimeUtils.cs
//Filename: DateTimeUtils.cs
//Version: 20131113

using System;

namespace ClipFlair.CaptionsLib.Utils
{
  
  public static class DateTimeUtils
  {

    #region --- Constants ---

    public static readonly DateTime DATETIMEZERO = new DateTime(1, 1, 1, 0, 0, 0, DateTimeKind.Local); //01/01/0001, 0:0:0.00 (dates must be from year 01 and above) '??? trying bigger year to avoid accuracy errors but it causes hichkups at the timebar

    #endregion

    #region --- Methods ---

    public static DateTime FixDateTime(DateTime datetime)
    {
      return new DateTime(DATETIMEZERO.Year, DATETIMEZERO.Month, DATETIMEZERO.Day, datetime.Hour, datetime.Minute, datetime.Second, datetime.Millisecond);
    }

    public static double DateTimeToSeconds(DateTime theDateTime, DateTime baseDateTime, int digits)
    {
      double result = (theDateTime - baseDateTime).TotalSeconds;
      if (digits != -1)
        return Math.Round(result, digits);
      return result;
    }

    public static DateTime SecondsToDateTime(double seconds, DateTime baseDateTime, int digits)
    {
      try {
        if (digits != -1)
          seconds = Math.Round(seconds, digits);
        return baseDateTime.AddSeconds(seconds);
      } catch {
        return baseDateTime;
        //silently handle exceptions, returning the baseDateTime in that case
      }
    }

    public static DateTime TimeStrToDateTime(string datetimeStr, DateTime baseDateTime, string datetimeFormat)
    {
      DateTime result;
      try {
        result = DateTime.ParseExact(datetimeStr, datetimeFormat, null);
      //if exact datetime parsing fails, try with the plain parse method
      } catch {
        result = DateTime.Parse(datetimeStr);
      }
      result = FixDateTime(baseDateTime + (result - baseDateTime));
      return result;
    }

    public static double TimeStrToSeconds(string datetimeStr, DateTime baseDateTime, string datetimeFormat, int digits)
    {
      return DateTimeToSeconds(TimeStrToDateTime(datetimeStr, baseDateTime, datetimeFormat), baseDateTime, digits);
    }

    public static string SecondsToDateTimeStr(double seconds, DateTime baseDateTime, string datetimeFormat, int digits)
    {
      return SecondsToDateTime(seconds, baseDateTime, digits).ToString(datetimeFormat);
    }

    #endregion

  }

}